Security/Http/Authenticator/OAuthAuthenticator.php

```diff
@@ -125,8 +120,40 @@ public function authenticate(Request $request): Passport
         $token = new OAuthToken($accessToken);
         $token->setResourceOwnerName($resourceOwner->getName());
 
+        return new SelfValidatedOAuthPassport($this->refreshToken($token));
+    }
+
+    /**
+     * This function can be used for refreshing an expired token
+     * or for custom "password grant" authenticator, if site owner also owns oauth instance.
+     *
+     * @template T of OAuthToken
+     *
+     * @param T $token
+     *
+     * @return T
+     */
+    public function refreshToken(OAuthToken $token): OAuthToken
+    {
+        $resourceOwner = $this->resourceOwnerMap->getResourceOwnerByName($token->getResourceOwnerName());
+
         if ($token->isExpired()) {
-            $token = $this->refreshToken($token, $resourceOwner);
+            $expiredToken = $token;
+            if ($refreshToken = $expiredToken->getRefreshToken()) {
+                $tokenClass = \get_class($expiredToken);
+                $token = new $tokenClass($resourceOwner->refreshAccessToken($refreshToken));
+                $token->setResourceOwnerName($expiredToken->getResourceOwnerName());
```

Comment:
I would add new method to `OAuthToken`, like:
```php
public method refreshToken(string $accessToken, OAuthToken $expiredToken): self
{
    $token = new self($accessToken);
    $token->setResourceOwnerName($expiredToken->getResourceOwnerName());
    $token->copyPersistentDataFrom($expiredToken);
    if (!$token->getRefreshToken()) {
        $token->setRefreshToken($expiredToken->getRefreshToken());
    }

    return $token;
}

Reply:
So after this it could be:
```php
$token = OAuthToken::refreshToken($resourceOwner->refreshAccessToken($refreshToken), $expiredToken);

Reply:
Hier is a problem. I do not know, whether OAuthToken or extended OAuthRememberMeToken will be just refreshed, more better hier to have an instance method:

```php
public function refresh(array $accessToken): self // statc
{
    $token = new static($accessToken);
    $token->setResourceOwnerName($this->getResourceOwnerName());
    $token->copyPersistentDataFrom($this);
    if (!$token->getRefreshToken()) {
        $token->setRefreshToken($this->getRefreshToken());
    }

    return $token;
}

// and we can call it on expiredToken
$expiredToken->refresh($resourceOwner->refreshAccessToken($refreshToken))
```
It must be guaranteed that function returns an instance of the same class.
Your example returns always an OAuthToken.

Reply:
`static` works for me.

Reply:
phpstan moans here: Unsafe usage of new static(). [https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static](https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static)

let us make this changes later if it becomes necessary.